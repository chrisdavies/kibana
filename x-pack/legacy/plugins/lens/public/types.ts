/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { Ast } from '@kbn/interpreter/common';
import { DragContextState } from './drag_drop';

// eslint-disable-next-line
export interface EditorFrameOptions {}

export type ErrorCallback = (e: { message: string }) => void;

export interface EditorFrameInstance {
  mount: (element: Element, props: { onError: ErrorCallback }) => void;
  unmount: () => void;
}

export interface EditorFrameSetup {
  createInstance: (options: EditorFrameOptions) => EditorFrameInstance;
  // generic type on the API functions to pull the "unknown vs. specific type" error into the implementation
  registerDatasource: <T, P>(name: string, datasource: Datasource<T, P>) => void;
  registerVisualization: <T, P>(name: string, visualization: Visualization<T, P>) => void;
}

// Hints the default nesting to the data source. 0 is the highest priority
export type DimensionPriority = 0 | 1 | 2;

// 0 is the default layer, layers are joined sequentially but there can only be one 'join' layer
export type DimensionLayer = 'join' | number;

export interface TableSuggestionColumn {
  columnId: string;
  operation: Operation;
  layer?: DimensionLayer;
}

export interface TableSuggestion {
  datasourceSuggestionId: number;
  isMultiRow: boolean;
  columns: TableSuggestionColumn[];
}

export interface DatasourceSuggestion<T = unknown> {
  state: T;
  table: TableSuggestion;
}

/**
 * Interface for the datasource registry
 */
export interface Datasource<T = unknown, P = unknown> {
  // For initializing, either from an empty state or from persisted state
  // Because this will be called at runtime, state might have a type of `any` and
  // datasources should validate their arguments
  initialize: (state?: P) => Promise<T>;

  // Given the current state, which parts should be saved?
  getPersistableState: (state: T) => P;

  insertLayer: (state: T, newLayerId: string) => T;

  renderDataPanel: (domElement: Element, props: DatasourceDataPanelProps<T>) => void;

  toExpression: (state: T) => Ast | string | null;

  getDatasourceSuggestionsForField: (state: T, field: unknown) => Array<DatasourceSuggestion<T>>;
  getDatasourceSuggestionsFromCurrentState: (state: T) => Array<DatasourceSuggestion<T>>;

  getPublicAPI: (state: T, setState: (newState: T) => void) => DatasourcePublicAPI;
}

/**
 * This is an API provided to visualizations by the frame, which calls the publicAPI on the datasource
 */
export interface DatasourcePublicAPI {
  getTableSpec: () => TableSpec;
  getOperationForColumnId: (columnId: string) => Operation | null;

  // Render can be called many times
  renderDimensionPanel: (domElement: Element, props: DatasourceDimensionPanelProps) => void;

  removeColumnInTableSpec: (columnId: string) => void;
  moveColumnTo: (columnId: string, targetIndex: number) => void;
  duplicateColumn: (columnId: string) => TableSpec;
}

export interface TableSpecColumn {
  // Column IDs are the keys for internal state in data sources and visualizations
  columnId: string;
  layer?: DimensionLayer;
}

// TableSpec is managed by visualizations
export type TableSpec = TableSpecColumn[];

export interface DatasourceDataPanelProps<T = unknown> {
  state: T;
  dragDropContext: DragContextState;
  setState: (newState: T) => void;
}

// The only way a visualization has to restrict the query building
export interface DatasourceDimensionPanelProps {
  // layerId: DimensionLayer;
  layerId: string;

  columnId: string;

  dragDropContext: DragContextState;

  // Visualizations can restrict operations based on their own rules
  filterOperations: (operation: Operation) => boolean;

  // Visualizations can hint at the role this dimension would play, which
  // affects the default ordering of the query
  suggestedPriority?: DimensionPriority;
}

export type DataType = 'string' | 'number' | 'date' | 'boolean';

// An operation represents a column in a table, not any information
// about how the column was created such as whether it is a sum or average.
// Visualizations are able to filter based on the output, not based on the
// underlying data
export interface Operation {
  // Operation ID is a reference to the operation
  id: string;
  // User-facing label for the operation
  label: string;
  // The output of this operation will have this data type
  dataType: DataType;
  // A bucketed operation is grouped by duplicate values, otherwise each row is
  // treated as unique
  isBucketed: boolean;

  // Extra meta-information like cardinality, color
}

// This is a temporary type definition, to be replaced with
// the official Kibana Datatable type definition.
export interface KibanaDatatable {
  type: 'kibana_datatable';
  rows: Array<Record<string, unknown>>;
  columns: Array<{ id: string; name: string }>;
}

export interface VisualizationProps<T = unknown> {
  dragDropContext: DragContextState;
  datasource: DatasourcePublicAPI;
  frame: FramePublicAPI;
  state: T;
  setState: (newState: T) => void;
}

export interface SuggestionRequest<T = unknown> {
  // It is up to the Visualization to rank these tables
  tables: TableSuggestion[];
  state?: T; // State is only passed if the visualization is active
}

export interface VisualizationSuggestion<T = unknown> {
  score: number;
  title: string;
  state: T;
  datasourceSuggestionId: number;
  previewExpression?: Ast | string;
  previewIcon: string;
}

export interface FramePublicAPI {
  datasourceLayers: Record<string, DatasourcePublicAPI>;
  layerIdToDatasource: Record<string, string>;
  // Adds a new layer. This triggers a re-render
  addNewLayer: () => void;
}

export interface Visualization<T = unknown, P = unknown> {
  // For initializing from saved object
  // initialize: (frame: FramePublicAPI, datasource: DatasourcePublicAPI, state?: P) => T;
  initialize: (frame: FramePublicAPI, state?: P) => T;

  getPersistableState: (state: T) => P;

  renderConfigPanel: (domElement: Element, props: VisualizationProps<T>) => void;

  // toExpression: (state: T, datasource: DatasourcePublicAPI) => Ast | string | null;
  toExpression: (state: T, frame: FramePublicAPI) => Ast | string | null;

  // The frame will call this function on all visualizations when the table changes, or when
  // rendering additional ways of using the data
  getSuggestions: (options: SuggestionRequest<T>) => Array<VisualizationSuggestion<T>>;
}
