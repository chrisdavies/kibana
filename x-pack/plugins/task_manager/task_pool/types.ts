/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

/**
 * This file contains all public type definitions, as well as any
 * definitions which need to be shared between modules. Internal types
 * are commented to indicate that they are not intended for use outside
 * of the task_pool module.
 */

// The Elasticsearch.js connection that is used to manage the
// task manager index.
export type CallCluster = (action: string, args: any) => Promise<any>;

// The Kibana logger
export type LogFn = (prefix: string[], message: string) => any;

/**
 * The object which is passed into a task's run function.
 *
 * @export
 * @interface TaskContext
 */
export interface TaskContext {
  /**
   * An elastic search connection that runs in the context of the
   * user (or system) that sheduled the task instance.
   *
   * @type {CallCluster}
   * @memberof TaskContext
   */
  callCluster: CallCluster;

  /**
   * An object containing the parameters that were specified when
   * the task was scheduled. The shape of this object is completely
   * up to the author of the task.
   *
   * @type {*}
   * @memberof TaskContext
   */
  params: any;

  /**
   * An object containing the result of the last successful run, if
   * this is a recurring task which has already succeeded one or more
   * times. If this is the first time the task has run, or if it has
   * never succeeded, this will be the empty object: {}
   *
   * @type {*}
   * @memberof TaskContext
   */
  previousResult: any;
}

/**
 * Defines a task which can be scheduled and run by the Kibana
 * task manager.
 */
export interface TaskDefinition {
  /**
   * A unique identifier for the type of task being defined.
   */
  type: string;

  /**
   * A brief, human-friendly title for this task.
   */
  title: string;

  /**
   * An optional more detailed description of what this task does.
   */
  description?: string;

  /**
   * A function which, does the work this task is built to do. Note,
   * this is a *function* and is not guaranteed to be called with
   * the *this* context of the task.
   *
   * @memberof TaskDefinition
   */
  run: (context: TaskContext) => Promise<any>;
}

export interface TaskDefinitions {
  [type: string]: TaskDefinition;
}

// Internal, a task that is currently running
export interface RunningTask {
  id: string;
  promise: Promise<any>;
  startTime: Date;
  type: string;
}

export type TaskStatus = 'idle' | 'running';

// Internal, more convenient representation of a task
export interface TaskDoc {
  attempts: number;
  id: string;
  interval?: string;
  nextRun: Date;
  params: object;
  previousResult: object;
  status: TaskStatus;
  timeOut: Date | null;
  type: string;
  version: number;
}

/**
 * Stats on the health of a task pool.
 */
export interface TaskPoolStats {
  /**
   * An integer representing how much space is left in the task
   * pool. (e.g. if a task pool's maxPoolSize is 10 and 3 tasks
   * are running, this will be 7).
   */
  freeCapacity: number;

  /**
   * Stats about individual tasks that are currently running
   * in the pool.
   */
  runningTasks: TaskStats[];
}

/**
 * Stats about an individual task taht is running in a pool
 */
export interface TaskStats {
  /**
   * The id of the running task.
   */
  id: string;

  /**
   * The type of the running task, corresponding to a task definition.
   */
  type: string;

  /**
   * The time the task started running in the pool.
   */
  startTime: Date;
}

// TaskStore is a light-weight layer on top of an elastic search
// connection. This is a raw, system-level connection that can make
// any changes neccessary to the index.
export interface TaskStore {
  availableTasks: (query: TaskQuery) => Promise<TaskDoc[]>;
  update: (task: TaskDoc) => Promise<TaskDoc>;
  remove: (id: string) => Promise<void>;
}

export interface TaskQuery {
  knownTypes: string[];
  runningIds: string[];
}
