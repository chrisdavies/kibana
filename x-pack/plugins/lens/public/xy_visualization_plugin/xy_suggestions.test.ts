/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { getSuggestions } from './xy_suggestions';
import { TableColumn } from '../types';

describe('xy_suggestions', () => {
  function numCol(columnId: string): TableColumn {
    return {
      columnId,
      operation: {
        dataType: 'number',
        id: `avg_${columnId}`,
        label: `Avg ${columnId}`,
        isBucketed: false,
      },
    };
  }

  function strCol(columnId: string): TableColumn {
    return {
      columnId,
      operation: {
        dataType: 'string',
        id: `terms_${columnId}`,
        label: `Top 5 ${columnId}`,
        isBucketed: true,
      },
    };
  }

  function dateCol(columnId: string): TableColumn {
    return {
      columnId,
      operation: {
        dataType: 'date',
        id: `date_histogram_${columnId}`,
        isBucketed: true,
        label: `${columnId} histogram`,
      },
    };
  }

  test('ignores invalid combinations', () => {
    expect(
      getSuggestions({
        tableColumns: {
          1: [dateCol('a')],
          2: [strCol('foo'), strCol('bar')],
        },
      })
    ).toEqual([]);
  });

  test('suggests a basic x y chart with date on x', () => {
    expect(
      getSuggestions({
        tableColumns: {
          0: [numCol('a')],
          1: [numCol('bytes'), dateCol('date')],
        },
      })
    ).toMatchSnapshot();
  });

  test('suggests a split x y chart with date on x', () => {
    expect(
      getSuggestions({
        tableColumns: {
          1: [numCol('price'), numCol('quantity'), dateCol('date'), strCol('product')],
        },
      })
    ).toMatchSnapshot();
  });

  test('supports multiple suggestions', () => {
    expect(
      getSuggestions({
        tableColumns: {
          1: [numCol('price'), dateCol('date')],
          2: [numCol('count'), strCol('country')],
        },
      })
    ).toMatchSnapshot();
  });

  test('handles two numeric values', () => {
    expect(
      getSuggestions({
        tableColumns: {
          1: [numCol('quantity'), numCol('price')],
        },
      })
    ).toMatchSnapshot();
  });

  test('places non-numeric columns on x', () => {
    expect(
      getSuggestions({
        tableColumns: {
          1: [
            numCol('num votes'),
            {
              columnId: 'mybool',
              operation: {
                dataType: 'boolean',
                id: 'mybool',
                isBucketed: false,
                label: 'Yes / No',
              },
            },
          ],
        },
      })
    ).toMatchSnapshot();
  });
});
