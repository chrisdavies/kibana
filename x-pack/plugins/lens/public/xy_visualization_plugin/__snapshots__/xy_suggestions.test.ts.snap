// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xy_suggestions handles two numeric values 1`] = `
Array [
  Object {
    "datasourceSuggestionId": "1",
    "score": 1,
    "state": Object {
      "legend": Object {
        "isVisible": true,
        "position": "right",
      },
      "seriesType": "bar",
      "splitSeriesAccessors": Array [],
      "stackAccessors": Array [],
      "title": "Avg price of Avg quantity",
      "x": Object {
        "accessor": "quantity",
        "position": "bottom",
        "showGridlines": false,
        "title": "Avg quantity",
      },
      "y": Object {
        "accessors": Array [
          "price",
        ],
        "position": "left",
        "showGridlines": false,
        "title": "Avg price",
      },
    },
    "title": "Avg price of Avg quantity",
  },
]
`;

exports[`xy_suggestions places non-numeric columns on x 1`] = `
Array [
  Object {
    "datasourceSuggestionId": "1",
    "score": 1,
    "state": Object {
      "legend": Object {
        "isVisible": true,
        "position": "right",
      },
      "seriesType": "bar",
      "splitSeriesAccessors": Array [],
      "stackAccessors": Array [],
      "title": "Avg num votes of Yes / No",
      "x": Object {
        "accessor": "mybool",
        "position": "bottom",
        "showGridlines": false,
        "title": "Yes / No",
      },
      "y": Object {
        "accessors": Array [
          "num votes",
        ],
        "position": "left",
        "showGridlines": false,
        "title": "Avg num votes",
      },
    },
    "title": "Avg num votes of Yes / No",
  },
]
`;

exports[`xy_suggestions suggests a basic x y chart with date on x 1`] = `
Array [
  Object {
    "datasourceSuggestionId": "1",
    "score": 1,
    "state": Object {
      "legend": Object {
        "isVisible": true,
        "position": "right",
      },
      "seriesType": "line",
      "splitSeriesAccessors": Array [],
      "stackAccessors": Array [],
      "title": "Avg bytes over date histogram",
      "x": Object {
        "accessor": "date",
        "position": "bottom",
        "showGridlines": false,
        "title": "date histogram",
      },
      "y": Object {
        "accessors": Array [
          "bytes",
        ],
        "position": "left",
        "showGridlines": false,
        "title": "Avg bytes",
      },
    },
    "title": "Avg bytes over date histogram",
  },
]
`;

exports[`xy_suggestions suggests a split x y chart with date on x 1`] = `
Array [
  Object {
    "datasourceSuggestionId": "1",
    "score": 1,
    "state": Object {
      "legend": Object {
        "isVisible": true,
        "position": "right",
      },
      "seriesType": "line",
      "splitSeriesAccessors": Array [
        "product",
      ],
      "stackAccessors": Array [],
      "title": "Avg price & Avg quantity over date histogram",
      "x": Object {
        "accessor": "date",
        "position": "bottom",
        "showGridlines": false,
        "title": "date histogram",
      },
      "y": Object {
        "accessors": Array [
          "price",
          "quantity",
        ],
        "position": "left",
        "showGridlines": false,
        "title": "Avg price & Avg quantity",
      },
    },
    "title": "Avg price & Avg quantity over date histogram",
  },
]
`;

exports[`xy_suggestions supports multiple suggestions 1`] = `
Array [
  Object {
    "datasourceSuggestionId": "1",
    "score": 1,
    "state": Object {
      "legend": Object {
        "isVisible": true,
        "position": "right",
      },
      "seriesType": "line",
      "splitSeriesAccessors": Array [],
      "stackAccessors": Array [],
      "title": "Avg price over date histogram",
      "x": Object {
        "accessor": "date",
        "position": "bottom",
        "showGridlines": false,
        "title": "date histogram",
      },
      "y": Object {
        "accessors": Array [
          "price",
        ],
        "position": "left",
        "showGridlines": false,
        "title": "Avg price",
      },
    },
    "title": "Avg price over date histogram",
  },
  Object {
    "datasourceSuggestionId": "2",
    "score": 1,
    "state": Object {
      "legend": Object {
        "isVisible": true,
        "position": "right",
      },
      "seriesType": "bar",
      "splitSeriesAccessors": Array [],
      "stackAccessors": Array [],
      "title": "Avg count of Top 5 country",
      "x": Object {
        "accessor": "country",
        "position": "bottom",
        "showGridlines": false,
        "title": "Top 5 country",
      },
      "y": Object {
        "accessors": Array [
          "count",
        ],
        "position": "left",
        "showGridlines": false,
        "title": "Avg count",
      },
    },
    "title": "Avg count of Top 5 country",
  },
]
`;
